rust/agama-server/src/web/common/issues.rs

```diff
@@ -0,0 +1,280 @@
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE LLC about this file by physical or electronic mail, you may
+// find current contact information at www.suse.com.
+
+//! Defines a service that keep tracks of the Agama issues.
+//!
+//! It is responsible for:
+//!
+//! * Querying the issues via D-Bus and keeping them in a cache.
+//! * Listening to D-Bus signals to keep the cache up-to-date.
+//! * Emitting `IssuesChanged` events, replacing
+//!   [issues_stream](crate::web::common::issues_stream).
+//!
+//! The following components are included:
+//!
+//! * [IssuesService] that runs on a separate task to hold the status.
+//! * [IssuesClient] that allows querying the [IssuesService] server about the
+//!   issues.
+//! * [IssuesRouter] which allows building a router, replacing
+//!   [issues_router](crate::web::common::issues_router).
+//!
+//! At this point, it only handles the issues that are exposed through D-Bus.
+
+use crate::web::EventsSender;
+use agama_lib::{http::Event, issue::Issue};
+use agama_utils::dbus::build_properties_changed_stream;
+use axum::{extract::State, routing::get, Json, Router};
+use std::collections::HashMap;
+use tokio::sync::{broadcast, mpsc, oneshot};
+use tokio_stream::StreamExt;
+use zbus::{
+    fdo::PropertiesChanged,
+    names::BusName,
+    zvariant::{Array, OwnedObjectPath},
+};
+
+type IssuesServiceResult<T> = Result<T, IssuesServiceError>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum IssuesServiceError {
+    #[error("Could not return the issues")]
+    SendIssues,
+    #[error("Could not get an answer from the service: {0}")]
+    RecvIssues(#[from] oneshot::error::RecvError),
+    #[error("Could not set the command")]
+    SendCommand,
+    #[error("Error parsing issues from D-Bus: {0}")]
+    InvalidIssue(#[from] zbus::zvariant::Error),
+    #[error("Error reading the issues: {0}")]
+    DBus(#[from] zbus::Error),
+    #[error("Invalid D-Bus name: {0}")]
+    DBusName(#[from] zbus::names::Error),
+    #[error("Could not send the event: {0}")]
+    SendEvent(#[from] broadcast::error::SendError<Event>),
+}
+
+#[derive(Debug)]
+enum IssuesCommand {
+    Get(String, String, oneshot::Sender<Vec<Issue>>),
+}
+
+/// Implements a Tokio task that holds the issues for each service.
+pub struct IssuesService {
+    issues: HashMap<String, Vec<Issue>>,
+    commands: mpsc::Receiver<IssuesCommand>,
+    events: EventsSender,
+    dbus: zbus::Connection,
+}
+
+impl IssuesService {
+    /// Sets up and starts the service as a Tokio task.
+    ///
+    /// Once it is started, the service waits for:
+    ///
+    /// * Commands from a client ([IssuesClient]).
+    /// * Relevant events from D-Bus.
+    pub async fn start(dbus: zbus::Connection, events: EventsSender) -> IssuesClient {
+        let (tx, rx) = mpsc::channel(16);
```

Comment:
why buffer size 16? For issues it looks quite big for me...for progress it would be probably too low.

Reply:
Yes, you are right, it is too big. I would reduce it to, e.g., 4.